// @sap/cds-runtime/lib/messaging/common-utils/AMQPClient.js
const ClientAmqp = require('@sap/xb-msg-amqp-v100').Client
const { connect, disconnect } = require('./connections')

const addDataListener = (client, queue, prefix, cb) =>
  new Promise((resolve, reject) => {
    const source = `${prefix}${queue}`
    client
      .receiver(queue)
      .attach(source)
      .on('data', async raw => {
        const buffer = Buffer.concat(raw.payload.chunks)
        const payload = JSON.parse(buffer.toString())
        const topic = raw.source.properties.to.replace(/^topic:\/*/, '')
        await cb(topic, payload, { done: raw.done, failed: raw.failed })
      })
      .on('subscribed', () => {
        resolve()
      })
  })

const sender = (client, optionsApp) => client.sender(`${optionsApp.appName}-${optionsApp.appID}`).attach('')

const emit = ({ data, event: topic, headers = {} }, sender, prefix) =>
  new Promise((resolve, reject) => {
    const message = { ...headers, data }
    const payload = { chunks: [Buffer.from(JSON.stringify(message))], type: 'application/json' }
    const msg = {
      done: resolve,
      failed: reject,
      payload,
      target: {
        properties: {
          to: `${prefix}${topic}`
        }
      }
    }
    sender.write(msg)
  })

class AMQPClient {
  constructor(optionsAMQP, optionsApp, prefix) {
    this.optionsAMQP = optionsAMQP
    this.optionsApp = optionsApp
    this.prefix = prefix
  }

  connect() {
    this.client = new ClientAmqp(this.optionsAMQP)
    this.sender = sender(this.client, this.optionsApp)
    return connect(this.client)
  }

  disconnect() {
    if (this.client) return disconnect(this.client)
  }

  emit(msg) {
    return emit(msg, this.sender, this.prefix.topic)
  }

  listen(queue, cb) {
    return addDataListener(this.client, queue, this.prefix.queue, cb)
  }
}

module.exports = AMQPClient
